app/celery/signals.py

```diff
@@ -0,0 +1,103 @@
+    task_success, task_failure, task_retry, task_prerun
+)
+from celery_singleton.util import generate_lock
+from celery_singleton.config import Config
+from celery_singleton.backends import get_backend
+from app.celery.worker import celery_app
+from app.database import SessionLocal
+from app.services import WebsocketManagerService, JobStatusService, InstructorService
+from app.models.job_status import JobStatusModel, JobStatus
+from app.schemas import JobSchema, WebsocketJobStatusMessage
+
+def handle_job_status_update(job_id: str, job_type: str | None, job_status: JobStatus) -> None:
+    """ Store a job status update in the database and dispatch a websocket event. """
+    async def _handle_job_status_update():
+        with SessionLocal() as session:
+            job_status_model = JobStatusService(session).create_job_status_update(job_id, job_type, job_status)
+            ws_message = WebsocketJobStatusMessage(
+                data=JobSchema.from_orm(job_status_model)
+            )
+            
+            if job_type == "downsync":
+                user_list = await InstructorService(session).list_instructors()
+            else:
+                user_list = []
+
+            WebsocketManagerService.publish_sync_pubsub_ws_message(ws_message, user_list)
+    
+    asyncio.run(_handle_job_status_update())
+
+@task_received.connect
+def task_received_handler(request, **kw):
+    handle_job_status_update(request.task_id, request.task_name, JobStatus.RECEIVED)
+
+@task_prerun.connect
+def task_prerun_handler(sender=None, task_id=None, **kw):
+    handle_job_status_update(task_id, sender.name, JobStatus.STARTED)
+
+@task_success.connect
+def task_success_handler(sender=None, result=None, **kw):
+    handle_job_status_update(sender.request.id, sender.name, JobStatus.SUCCESS)
+
+@task_failure.connect
+def task_failure_handler(sender=None, task_id=None, **kw):
+    handle_job_status_update(task_id, sender.name, JobStatus.FAILURE)
+
+@task_internal_error.connect
+def task_internal_error_handler(sender=None, task_id=None, **kw):
+    handle_job_status_update(task_id, sender.name, JobStatus.FAILURE)
+
+@task_retry.connect
+def task_retry_handler(sender=None, request=None, **kw):
+    handle_job_status_update(request.task_id, request.task_name, JobStatus.RETRY)
+
+@task_revoked.connect
+def task_revoked_handler(sender=None, request=None, **kw):
+    handle_job_status_update(request.task_id, request.task_name, JobStatus.REVOKED)
+
+""" Source: https://github.com/steinitzu/celery-singleton/issues/29 """
+@task_revoked.connect
+def clean_singleton_lock_on_revoke(sender=None, request=None, reason=None, **kwargs):
+    """ For some reason, Celery devs decided to remove basically every task workflow callback,
+    including on_revoked, which means that celery-singleton has no idea if a task is revoked.
+    We need to manally fix this here...
+    """
+    task_id = request.id
+    task_name = request.task
```

Comment:
the `request` parameter is a `celery.app.task.Context` object, not a `task.request`, so I don't think it has a `task` field. here's the class definition for `Context`: https://docs.celeryq.dev/en/latest/_modules/celery/app/task.html#Context

Reply:
I recall looking into this some when originally implementing it, because this note in the documentation also caught my attention. Although it looks like I made a mistake anyways because as the documentation states, when using prefork pooling, the `task` field will not be defined. Perhaps this is a restriction we have to live with (no preforking), since we certainly need access to the underlying task function and I'm not sure how we would otherwise access it.